### PR TITLE
feat: add json-data type support in redis

### PR DESCRIPTION
**Pull request by sanpj2292**
# Description

Through this PR, we are planning to support JSON data-type in Redis

## Linear Ticket

Resolves INT-2095

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: utils/tests/redis.go
Comment (achettyiitr) [inline utils/tests/redis.go:1]: Can we use it from the go-kit, instead of this?
https://github.com/rudderlabs/rudder-go-kit/blob/main/testhelper/docker/resource/redis/redis.go

Comment (achettyiitr) [reply utils/tests/redis.go:1]: We would need a small change around adding WithRepository function option for overriding the repository setting.

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 1]
```diff
@@ -197,3 +269,154 @@ func TestKVManagerInvocations(t *testing.T) {
...
+
+	for _, tc := range redisJSONTestCases {
+		t.Run(tc.description, func(t *testing.T) {
+			redisAddr, destroy := testutils.StartRedis(t, tc.redisImgRepo, tc.redisImgTag)
+			defer destroy()
+			event, err := json.Marshal(tc.transformedResponse)
+			if err != nil {
+				t.Errorf("MarshalError: %v\n", err.Error())
+				return
+			}
```
Comment (achettyiitr) [inline router/customdestinationmanager/customdestinationmanager_test.go:293]: Less code plus internally `t.Errorf` calls `Fail()` only.
```suggestion
			require.NoError(t, err)
```

Comment (achettyiitr) [reply router/customdestinationmanager/customdestinationmanager_test.go:293]: Probably for other instances as well.

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 2]
```diff
@@ -197,3 +269,154 @@ func TestKVManagerInvocations(t *testing.T) {
...
+			Config: config,
+		}
+		kvMgr.Connect()
+		db := kvMgr.GetClient()
+
+		stCd, _ := customManager.send(event, kvMgr, config)
+		v, err := db.JSONGet(context.TODO(), "user:myuser-id", "$.mode-1").Result()
+
+		require.Nil(t, err)
+		require.NotEqual(t, "", v)
```
Comment (achettyiitr) [inline router/customdestinationmanager/customdestinationmanager_test.go:377]: We don't need this assertion, since below we are verifying for a expected string.

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 3]
```diff
@@ -197,3 +269,154 @@ func TestKVManagerInvocations(t *testing.T) {
...
+		}
+		kvMgr := &kvstoremanager.RedisManagerT{
+			Config: config,
+		}
+		kvMgr.Connect()
+		db := kvMgr.GetClient()
+		ctx := context.TODO()
+
+		_, e := db.JSONSet(ctx, "user:myuser-id", "$", `{"mode-in":{"a":1,"size":"LM"}}`).Result()
+		require.Nil(t, e)
```
Comment (achettyiitr) [inline router/customdestinationmanager/customdestinationmanager_test.go:406]: `err` seems more appropriate here.
```suggestion
		_, err = db.JSONSet(ctx, "user:myuser-id", "$", `{"mode-in":{"a":1,"size":"LM"}}`).Result()
		require.Nil(t, err)
```

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 4]
```diff
@@ -197,3 +269,154 @@ func TestKVManagerInvocations(t *testing.T) {
...
+		require.NotEqual(t, "", firstVal)
+
+		v, err2 := db.JSONGet(ctx, "user:myuser-id", "$.mode-1").Result()
+		require.Nil(t, err2)
+		require.NotEqual(t, "", v)
+
+		require.JSONEq(t, `[{"a":1,"size":"LM"}]`, firstVal)
+		require.JSONEq(t, `[{"key":"someKey","fields":{"field1":"value1","field2":2}}]`, v)
+
+		require.Equal(t, http.StatusOK, stCd)
```
Comment (achettyiitr) [inline router/customdestinationmanager/customdestinationmanager_test.go:420]: We should keep the assertions where it is used,
```suggestion
		stCd, _ := customManager.send(event, kvMgr, c)
		require.Equal(t, http.StatusOK, stCd)

		firstVal, err := db.JSONGet(ctx, "user:myuser-id", "$.mode-in").Result()
		require.NoError(t, err)
		require.JSONEq(t, `[{"a":1,"size":"LM"}]`, firstVal)

		v, err := db.JSONGet(ctx, "user:myuser-id", "$.mode-1").Result()
		require.NoError(t, err)
		require.JSONEq(t, `[{"key":"someKey","fields":{"field1":"value1","field2":2}}]`, v)
```

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 5]
```diff
@@ -197,3 +269,154 @@ func TestKVManagerInvocations(t *testing.T) {
...
+
+		stCd, _ := customManager.send(event, kvMgr, config)
+		v, err := db.JSONGet(context.TODO(), "user:myuser-id", "$.mode-1").Result()
+
+		require.Nil(t, err)
+		require.NotEqual(t, "", v)
+
+		require.JSONEq(t, `[{"key":"someKey","fields":{"field1":"value1","field2":2}}]`, v)
+
+		require.Equal(t, http.StatusOK, stCd)
```
Comment (achettyiitr) [inline router/customdestinationmanager/customdestinationmanager_test.go:381]: Same as below.
```suggestion
		stCd, _ := customManager.send(event, kvMgr, c)
		require.Equal(t, http.StatusOK, stCd)
		
		v, err := db.JSONGet(context.TODO(), "user:myuser-id", "$.mode-1").Result()
		require.Nil(t, err)
		require.JSONEq(t, `[{"key":"someKey","fields":{"field1":"value1","field2":2}}]`, v)
```

File: services/kvstoremanager/redis.go [hunk 6]
```diff
@@ -26,22 +29,35 @@ func init() {
...
+	redis.Cmdable
+}
+
+func (m *RedisManagerT) GetClient() redisCli {
+	var redisClient redisCli = m.client
+	if m.clusterMode {
+		redisClient = m.clusterClient
+	}
+	return redisClient
+}
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply services/kvstoremanager/redis.go]: Makes sense. Refactored.

File: services/kvstoremanager/redis.go [hunk 7]
```diff
@@ -4,20 +4,23 @@ import (
...
 	"github.com/rudderlabs/rudder-server/utils/types"
 )
 
 var abortableErrors = []string{}
 
-type redisManagerT struct {
+type RedisManagerT struct {
 	clusterMode   bool
-	config        types.ConfigT
+	Config        types.ConfigT
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply services/kvstoremanager/redis.go]: Since `NewRedisManager` method was not present, we exposed `Config`. But this has been refactored now. This is done

File: services/kvstoremanager/redis.go [hunk 8]
```diff
@@ -4,20 +4,23 @@ import (
...
 	"github.com/redis/go-redis/v9"
+	"github.com/tidwall/gjson"
 
 	"github.com/rudderlabs/rudder-server/utils/types"
 )
 
 var abortableErrors = []string{}
 
-type redisManagerT struct {
+type RedisManagerT struct {
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply services/kvstoremanager/redis.go:21]: Done

File: services/kvstoremanager/redis.go [hunk 9]
```diff
@@ -150,3 +166,65 @@ func (m *redisManagerT) HSet(hash, key string, value interface{}) (err error) {
...
 }
+
+func (m *RedisManagerT) ExtractJSONSetArgs(jsonData json.RawMessage) ([]string, error) {
+	key := gjson.GetBytes(jsonData, "message.key").String()
+	path := gjson.GetBytes(jsonData, "message.path").String()
+	jsonVal := gjson.GetBytes(jsonData, "message.value")
+	var redisClient redisCli = m.client
+	if m.clusterMode {
+		redisClient = m.clusterClient
+	}
```
[the remark this answers is not in the document]
Comment (sanpj2292) [reply services/kvstoremanager/redis.go]: Done

File: router/customdestinationmanager/customdestinationmanager_test.go (hunk 1 above)
Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go]: Done

File: router/customdestinationmanager/customdestinationmanager_test.go (hunk 3 above)
Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go]: Updated to a more appropriate one

File: router/customdestinationmanager/customdestinationmanager_test.go (hunk 5 above)
Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go]: Done

File: router/customdestinationmanager/customdestinationmanager_test.go (hunk 4 above)
Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go]: Done

File: router/customdestinationmanager/customdestinationmanager_test.go [hunk 10]
```diff
@@ -197,3 +269,126 @@ func TestKVManagerInvocations(t *testing.T) {
...
+		require.NoError(t, err)
+		config := map[string]interface{}{
+			"useJSONModule": true,
+			"address":       redisAddr,
+			"db":            0,
+			"clusterMode":   false,
+		}
+		kvMgr := kvstoremanager.NewRedisManager(config)
+		db := kvMgr.GetClient()
+		ctx := context.TODO()
```
Comment (fracasula) [inline router/customdestinationmanager/customdestinationmanager_test.go:378]: `context.TODO` is used to mark a contenxt that you are planning to replace in the future with a different one (e.g. one with a timeout). Is that what you are planning TODO (pun intended)? :stuck_out_tongue_winking_eye:

Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go:378]: We can think that way 😉 . On a serious note, Empty context is the intention here 🙂

File: services/kvstoremanager/redis.go [hunk 11]
```diff
@@ -150,3 +168,60 @@ func (m *redisManagerT) HSet(hash, key string, value interface{}) (err error) {
...
+
+func (*RedisManager) ShouldSendDataAsJSON(config map[string]interface{}) bool {
+	var dataAsJSON bool
+	if dataAsJSONI, ok := config["useJSONModule"]; ok {
+		if dataAsJSON, ok = dataAsJSONI.(bool); ok {
+			return dataAsJSON
+		}
+	}
+	return dataAsJSON
+}
```
Comment (fracasula) [inline services/kvstoremanager/redis.go:227]: Why is this in the `RedisManager`? It doesn't seem to be related to Redis operations specifically. In fact it does not need a client and there is no struct receiver. The only dependency is the configuration which is passed as a parameter.

I suggest moving this elsewhere. The custom destination manager seems to be a good candidate. Thoughts?

File: router/customdestinationmanager/customdestinationmanager_test.go (hunk 10 above)
Comment (fracasula) [reply router/customdestinationmanager/customdestinationmanager_test.go:378]: `context.Background()` it is then :+1: :slightly_smiling_face:

Comment (sanpj2292) [reply router/customdestinationmanager/customdestinationmanager_test.go]: Done

File: utils/tests/redis.go
Comment (achettyiitr) [inline utils/tests/redis.go:1]: We have got the release tag for https://github.com/rudderlabs/rudder-go-kit/releases/tag/v0.31.0 with the corresponding changes for the custom Redis docker repository. We can now get rid of this file and use that.

Comment (sanpj2292) [reply utils/tests/redis.go:1]: This is done

**Review by fracasula — COMMENTED** [review]
Approved with minor comment. It should be fairly easy to do if you agree.

File: services/kvstoremanager/redis.go [hunk 12]
```diff
@@ -26,7 +29,22 @@ func init() {
...
+
+func (m *RedisManager) GetClient() redis.Cmdable {
+	if m.clusterMode {
+		return m.clusterClient
+	}
+	return m.client
+}
+
+func (m *RedisManager) Connect() {
```
Comment (fracasula) [inline services/kvstoremanager/redis.go:47]: This method doesn't connect to anything, in fact it doesn't return an error. If it would connect it would have to return one. I think we should rename it to `CreateClient()`.

**Review by fracasula — APPROVED** [review]
Approved with a minor comment. It _**should**_ be fairly easy to do if you agree.

File: services/kvstoremanager/redis.go (hunk 12 above)
Comment (sanpj2292) [reply services/kvstoremanager/redis.go]: Makes sense

Comment (sanpj2292) [reply services/kvstoremanager/redis.go]: Changes are done

**Review by fracasula — APPROVED** [review]

**Review by achettyiitr — APPROVED** [review]

File: services/kvstoremanager/redis.go (hunk 11 above)
Comment (koladilip) [inline services/kvstoremanager/redis.go:227]:
```suggestion
func (*RedisManager) ShouldSendDataAsJSON(config map[string]interface{}) bool {
	if dataAsJSON, ok := config["useJSONModule"]; ok {
		if dataAsJSON, ok = dataAsJSON.(bool); ok {
			return dataAsJSON
		}
	}
	return false
}
```

Comment (koladilip) [reply services/kvstoremanager/redis.go:227]: optional

**Review by koladilip — APPROVED** [review]

File: services/kvstoremanager/redis.go [hunk 13]
```diff
@@ -150,3 +168,60 @@ func (m *redisManagerT) HSet(hash, key string, value interface{}) (err error) {
...
+	path := gjson.GetBytes(jsonData, "message.path").String()
+	jsonVal := gjson.GetBytes(jsonData, "message.value")
+
+	actualPath := "$" // root insert
+	if path != "" {
+		actualPath = fmt.Sprintf("$.%s", path)
+	}
+	args := []string{key, actualPath, jsonVal.String()}
+
+	if actualPath != "$" {
```
Comment (koladilip) [inline services/kvstoremanager/redis.go:183]: can you add some comments here

**Review by koladilip — APPROVED** [review]

File: services/kvstoremanager/redis.go (hunk 13 above)
Comment (sanpj2292) [reply services/kvstoremanager/redis.go:195]: Added them. Can you check and let me know if they are ok ?

File: services/kvstoremanager/redis.go (hunk 11 above)
Comment (sanpj2292) [reply services/kvstoremanager/redis.go:239]: Skipping this for now.
